File: 프로그래머스/lv1/72410. 신규 아이디 추천/신규 아이디 추천.py
import re


def solution(new_id):
    # 1
    new_id = new_id.lower()
    # 2
    new_id = re.sub("[^a-z0-9-_.]", '', new_id)
    # 3
    new_id = re.sub('\\.{2,}', '.', new_id)
    # 4
    if new_id[0] == '.':
        new_id = new_id[1:]
    if len(new_id) > 0 and new_id[-1] == '.':
        new_id = new_id[:-1]
    # 5
    if new_id == "":
        new_id = "a"
    # 6
    if len(new_id) >= 16:
        new_id = new_id[:15]
        if new_id[-1] == '.':
            new_id = new_id[:-1]
    # 7
    while len(new_id) <= 2:
        new_id = new_id + new_id[-1]
    return new_id

File: 프로그래머스/lv1/72410. 신규 아이디 추천/신규 아이디 추천.java
class Solution {
    public String solution(String newId) {
        String answer = newId.toLowerCase()
            .replaceAll("[^a-z0-9\\-_.]", "")
            .replaceAll("\\.+", ".")
            .replaceAll("^\\.|\\.$", "");
        if(answer.length() > 15){
            answer = answer.substring(0, 15)
            .replaceAll("^\\.|\\.$", "");
        }
        if(answer.length() == 0){
            answer = "aaa";
        }
        while(answer.length() <= 2){
            answer += answer.substring(answer.length() - 1, answer.length());
        }
        
        return answer;
    }
}